File: solutions/Balanced-Binary-Tree/Balanced-Binary-Tree.cpp

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalancedHelper(TreeNode *root, int &depth)
    {
        if(root==NULL)
        {
            depth=0;
            return true;
        }
        int left,right;
        if(isBalancedHelper(root->left, left) && isBalancedHelper(root->right, right))
        {
            int diff=abs(left-right);
            if(diff<=1)
            {
                depth=1+(left>right?left:right);
                return true;
            }
        }
        return false;
    }
    bool isBalanced(TreeNode *root) {
        int depth=0;
        return isBalancedHelper(root, depth);
    }
};


 

File: solutions/Balanced-Binary-Tree/Balanced-Binary-Tree.js
const dfsHeight = (root) => {
  if (root === null) {
    return 0;
  }
  const left = dfsHeight(root.left);
  if (left === -1) {
    return -1;
  }
  const right = dfsHeight(root.right);
  if (right === -1) {
    return -1;
  }
  return Math.abs(left - right) > 1 ? -1 : Math.max(left, right) + 1;
};

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function (root) {
  return dfsHeight(root) !== -1;
};